### PR TITLE
Migrate to CircleCI

### PR DESCRIPTION
**Pull request by mdeggies**
This PR migrates the tests (bundle install & bundle exec rake) from Travis -> Circle, and moves the logic from TeamCity -> Circle that does the docker image push & rubygems push on merges to master. A cache is used & restored when the gemfile hasn't changed which should speed up tests, and a context is used to store the environment variables that are needed to do the rubygems/dockerhub pushes. 

To keep the environment the same as it is in Travis, I explicitly set the bundle and gem versions to those that are used in the latest Travis runs. These can be bumped up as needed.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+          gem-version: '3.1.2'
+      - save_cache: *save_gem_cache
+      - persist_to_workspace:
+          root: .
+          paths:
+            - vendor/bundle
+  gem-push:
+    docker:
+      - image: *RUBY_IMAGE
+    working_directory: ~/data
```
Comment (alvin-huang) [inline .circleci/config.yml:67]: Do you need to set a `working_directory` here? Or can it just use the default?

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+workflows:
+  middleman-hashicorp:
+    jobs:
+      - build
+      - gem-push:
+          context: middleman-hashicorp
+          requires: 
+            - build
+          filters:
```
Comment (alvin-huang) [inline .circleci/config.yml:101]: In TeamCity the builds only happened on tags and this is [possible](https://circleci.com/docs/2.0/configuration-reference/#tags) in Circle too. Ex:

```
          filters:
            tags:
              only: /^v\d+\.\d+\.\d+$/
```

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            branches:
+              only:
+                - master
+      - docker-push:
+          context: middleman-hashicorp
+          requires: 
+            - build
+          filters:
```
Comment (alvin-huang) [inline .circleci/config.yml:109]: Same here

File: .circleci/config.yml (hunk 2 above)
Comment (mdeggies) [reply .circleci/config.yml:101]: Thanks for catching this alvin!

File: .circleci/config.yml (hunk 1 above)
Comment (mdeggies) [reply .circleci/config.yml:62]: I changed this around a bit after testing. Calling `bundle exec rake release` from ~/data works without needing any other commands, when the contents of ~/data/vendor/bundle are saved/restored from the previous step

Comment (mdeggies) [reply .circleci/config.yml:54]: Let me know how this looks now, and thanks again for the review, Alvin!

**Pull request merged**